### ISSUE TITLE
Decide what to test next

### ISSUE THREAD

**Issue by danwt** [cosmos/interchain-security#335]
## Testing Priorities

We need to decide what to test next. Particularly for the goal of finding deeper bugs. (AKA not unit testing).

### Problem

A lot of important pieces of code are still not tested thoroughly enough (see [Quality Assurance Plan](https://github.com/cosmos/interchain-security/blob/main/docs/quality_assurance.md)). We need to decide the next testing efforts.

We have unit tests, handmade e2e tests, trace based e2e tests (difference tests), and e2e integration tests.

Unit tests can't find more complicated logic bugs. Handmade e2e tests can, but require a lot of upfront and maintenance effort, and give little confidence that something hasn't been missed. The e2e integration tests test a lot of code, but the scenarios that can be tested with them are limited and they are slow ect. Trace based tests can be a good way to achieve a high degree of coverage, but require maintenance of a model and creating a mapper trace -> test (the driver).

I think we would still benefit from adding more unit tests, but this issue is about how we can find more complicated logic bugs.

### Description

We need to decide what to test next before we can decide how to test it. 

In my view there are poor candidates and good candidates for what to focus on. The poor candidates are poor because they're either already tested, or because their incorrectness wouldn't matter as much. The good candidates are good because they are untested, crucial for correctness, and are 'low hanging fruit' in terms of how easy it would be to do the work to test them. 

#### Poor candidates

I think it would be a poor use of time to test the following **next**

1. Anything in ccv/consumer.
Mainly because it seems higher priority to test the provider. Also difference/core tests that the consumer receives VSC packets and sends maturities at appropriate times. It also tests that consumers send slashes and receive slash acknowledgements appropriately. 
2. Sending VSC packets from provider to a single consumer.\
difference/core test this with 1 consumer.
3. Receiving Maturity packets on provider from a single consumer.\
difference/core test this with 1 consumer.
4. Validator power change on provider.\
difference/core tests this e2e style starting from delegate/undelegate/jail ops.
5. Consumer initiated slashes on provider from 1 consumer.\
difference/core tests this with 0.00 slash factors, which shouldn't make any difference as positive slash factors just burn tokens. (4) has the same effect.
6. Provider initiated slashing.\
I will double check but I don't see any reason to test this as it is just a jail + token burn action which are already tested by (4, 5).
7. Anything related to staking module API.\
difference/core tests that validators and a delegator always have consistent token balances as per the [Bond Based Consumer Voting Power](https://github.com/cosmos/ibc/blob/main/spec/app/ics-028-cross-chain-validation/system_model_and_properties.md#system-properties) property.
8. Unbonding occurs neither early or late with 1 consumer.\
difference/core tests this via (2,3,4,7).

#### Good candidates

I think it could be a good use of time to test the following next

1. When a provider interleaves CreateConsumer, handshake and StopConsumer steps for multiple consumers, unbondingCanComplete calls occur neither too early nor too late.\
In my view this is likely the crucial property to test next, as combined with 'Poor candidates'::(8) this would give good confidence.
2. (1) except for with a more abstract handshake.\
We could collapse handshake steps, to not interleavings of every step. This might be easier than (1).

### How?

Once we decide on exactly what to test, we should think about how to do it. New issues can be created for that.

### Resources

- [Quality Assurance Plan](https://github.com/cosmos/interchain-security/blob/main/docs/quality_assurance.md)

## Closing criteria for this issue

This issue should be closed when we know exactly what we want to get out of next testing efforts.

**Comment by danwt** [comment]
Here is an issue about _how_ we might test scenarios 'Good candidates'::(1,2) if we wanted to do that

- https://github.com/cosmos/interchain-security/issues/336

**Comment by jtremback** [comment]
> Any consumer behavior. difference/core tests that the consumer receives VSC packets and sends maturities at appropriate times. It also tests that consumers send slashes and receive slash acknowledgements appropriately.

What do you mean by "Any consumer behavior" in this case?

> When a provider interleaves CreateConsumer, handshake and StopConsumer steps for multiple consumers, unbondingCanComplete calls occur neither too early nor too late.
In my view this is likely the crucial property to test next, as combined with 'Poor candidates'::(8) this would give good confidence.

Is it important to also test how pending packets of various types behave? Also, a concern I have is whether it is possible to restart both chains at every step of the handshake. Maybe this is better to do in integration tests.

**Comment by danwt** [comment]
> What do you mean by "Any consumer behavior" in this case?

Hey, good point, that wasn't a well written bullet. I thought that we mostly care about making sure the provider is working and robust right now, so it should take priority. 

> Is it important to also test how pending packets of various types behave?

For the pendingVSC's from provider to consumer definitely yes. I think pending slashes from consumer to provider are low priority.

> Also, a concern I have is whether it is possible to restart both chains at every step of the handshake. Maybe this is better to do in integration tests.

Ok, noted! The point of this issue is to collect these ideas.

On top of this ^^, and related, is various handshake crossovers. @mpoke is currently looking into this AFAIK

**Comment by danwt** [comment]
Closing as out of date - I'll make a new writeup for what will be done next.
